### ISSUE TITLE
NeedData exception on a valid pcap

### ISSUE THREAD

**Issue by danielzyp** [kbandla/dpkt#273]
```
for pcap_path in pcaps:
    f = open(pcap_path)
    pcap = dpkt.pcap.Reader(f)
    i = 1
    for ts, buf in pcap:
        eth = dpkt.ethernet.Ethernet(buf)
        ip = eth.data
        tcp = ip.data

        if tcp.dport == 80 and len(tcp.data) > 0:
            http = dpkt.http.Request(tcp.data)
            lst.append(http.headers['host'])
    f.close()
```

Throws NeedData exception on the following packets:

```
Failed in packet 1
Failed in packet 6
Failed in packet 7
Failed in packet 8
Failed in packet 10
Failed in packet 12
Failed in packet 14
Failed in packet 19
Failed in packet 21
Failed in packet 22
Failed in packet 24
Failed in packet 26
Failed in packet 28
Failed in packet 30
Failed in packet 32
Failed in packet 34
Failed in packet 36
Failed in packet 38
Failed in packet 41
Failed in packet 42
Failed in packet 45
Failed in packet 46
Failed in packet 48
Failed in packet 50
Failed in packet 52
Failed in packet 54
Failed in packet 57
Failed in packet 58
Failed in packet 60
Failed in packet 62
Failed in packet 64
Failed in packet 68
Failed in packet 70
Failed in packet 72
Failed in packet 78
Failed in packet 80
Failed in packet 90
Failed in packet 92
Failed in packet 94
Failed in packet 98
Failed in packet 100
Failed in packet 102
Failed in packet 106
Failed in packet 108
Failed in packet 110
Failed in packet 114
Failed in packet 116
Failed in packet 118
Failed in packet 120
Failed in packet 124
Failed in packet 126
Failed in packet 128
Failed in packet 130
Failed in packet 132
Failed in packet 134
Failed in packet 137
Failed in packet 143
Failed in packet 145
Failed in packet 155
Failed in packet 157
Failed in packet 159
Failed in packet 161
Failed in packet 163
Failed in packet 165
Failed in packet 169
Failed in packet 171
Failed in packet 173
Failed in packet 175
Failed in packet 178
Failed in packet 180
Failed in packet 184
Failed in packet 186
Failed in packet 188
Failed in packet 190
Failed in packet 193
Failed in packet 194
Failed in packet 196
Failed in packet 200
Failed in packet 202
Failed in packet 204
Failed in packet 208
Failed in packet 210
Failed in packet 212
Failed in packet 216
Failed in packet 218
Failed in packet 220
Failed in packet 226
Failed in packet 228
Failed in packet 238
Failed in packet 240
Failed in packet 242
Failed in packet 244
Failed in packet 248
Failed in packet 250
Failed in packet 252
Failed in packet 256
Failed in packet 258
Failed in packet 260
Failed in packet 264
Failed in packet 266
Failed in packet 268
Failed in packet 272
Failed in packet 274
Failed in packet 276
Failed in packet 280
Failed in packet 282
Failed in packet 284
Failed in packet 288
Failed in packet 290
Failed in packet 292
Failed in packet 296
Failed in packet 298
Failed in packet 300
Failed in packet 304
Failed in packet 306
Failed in packet 308
Failed in packet 312
Failed in packet 314
Failed in packet 316
```

And [here is the pcap](http://s000.tinyupload.com/index.php?file_id=55290766627538578957)

**Comment by kbandla** [comment]
TL;DR: I understand what you are trying to do. It should be easy to get the http request host header. In the interim, you can make this work my commenting [this](https://github.com/kbandla/dpkt/blob/master/dpkt/http.py#L64) line, and adding a dummy `pass` statement in its stead. 

@danielzyp , a few things were happening:
* the pcap has incomplete ethernet packets
* not all the packets are tcp
* outgoing http data is malware, might not always in the correct http format

That said, you should:
* first look for ethernet packets that are of type `ip`
* then, look for ip packets that have a protocol of `0x06` (tcp)
* reassemble tcp streams if needed (for http responses)

here is some quick code that does some of these checks: (not tested)
```
import dpkt

f = open('t3.pcap')
pcap = dpkt.pcap.Reader(f)
for num, (ts, buf) in enumerate(pcap):
    try:
        eth = dpkt.ethernet.Ethernet(buf)
    except dpkt.NeedData as e:
        print 'Error parsing packet #', num
        continue

    # ignore if the packet is not an IP packet
    if not eth.type == dpkt.ethernet.ETH_TYPE_IP:
        continue

    ip = eth.data
    # ignore if not TCP
    if not ip.p == dpkt.ip.IP_PROTO_TCP:
        continue

    tcp = ip.data
    try:
        if tcp.dport == 80 and len(tcp.data) > 0:
            http = dpkt.http.Request(tcp.data)
            host_header = http.headers['host']
            print host_header
    except Exception as e:
        print 'Error : %s'%e
f.close()
```
dpkt reads the http content-length header and expects the length of the data to match it. We probably should make it go easy on it. I am going to change the issue name to reflect the change needed.

**Comment by danielzyp** [comment]
Awesome :)
Works great
Also it's like x150 faster than scapy for some reason

**Comment by danielzyp** [comment]
It still fails sometimes, for example [this pcap](http://s000.tinyupload.com/index.php?file_id=65150854922959575810)

**Comment by kbandla** [comment]
@danielzyp no `tcp.dport==80` packets in that pcap
